cpp/include/raft/linalg/mdarray.h

```diff
@@ -0,0 +1,389 @@
+ *   CUDA stream is honorred.
+ *
+ * - For container policy, we require an additional view method for it to be used in
+ *   mdspan.  The container policy reference implementation has multiple `access` methods
+ *   that accommodate needs for both mdarray and mdspan.  This is more difficult for us
+ *   since the policy might contain states that are unwanted inside a CUDA kernel.  Also,
+ *   on host we return a proxy to the actual value as `device_ref` so different access
+ *   methods will have different return type, which is less desirable.
+ */
+template <class ElementType, class Extents, class LayoutPolicy, class ContainerPolicy>
+class mdarray {
+ public:
+  using element_type = ElementType;
+  using extents_type = Extents;
+  using layout_type  = LayoutPolicy;
+  using mapping_type = typename layout_type::template mapping<extents_type>;
+
+  using index_type            = std::size_t;
+  using size_type             = std::size_t;
+  using difference_type       = std::ptrdiff_t;
+  using container_policy_type = ContainerPolicy;
+  using container_type        = typename container_policy_type::container_type;
+
+  static_assert(!std::is_const<ElementType>::value,
+                "Element type for container must not be const.");
+
+  using pointer         = typename container_policy_type::pointer;
+  using const_pointer   = typename container_policy_type::const_pointer;
+  using reference       = typename container_policy_type::reference;
+  using const_reference = typename container_policy_type::const_reference;
+
+ private:
+  template <typename E>
+  using view_type_impl = std::conditional_t<
+    container_policy_type::is_host_type::value,
+    host_mdspan<E, extents_type, layout_type, typename container_policy_type::accessor_policy>,
+    device_mdspan<E, extents_type, layout_type, typename container_policy_type::accessor_policy>>;
+
+
+ public:
+  /**
+   * \brief the mdspan type returned by view method.
+   */
+  using view_type = view_type_impl<element_type>;
+  using const_view_type = view_type_impl<element_type const>;
+
+ public:
+  constexpr mdarray() noexcept(std::is_nothrow_default_constructible<container_type>::value) =
+    default;
+  constexpr mdarray(mdarray const&) noexcept(
+    std::is_nothrow_copy_constructible<container_type>::value) = default;
+  constexpr mdarray(mdarray&&) noexcept(std::is_nothrow_move_constructible<container_type>::value) =
+    default;
+
+  auto operator =(mdarray&&) noexcept(std::is_nothrow_move_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  auto operator =(mdarray const&) noexcept(std::is_nothrow_copy_assignable<container_type>::value)
+    -> mdarray& = default;
+
+  ~mdarray() noexcept(std::is_nothrow_destructible<container_type>::value) = default;
+
+  /**
+   * @brief The only constructor that can create storage, this is to make sure CUDA stream is being
+   * used.
+   */
+  constexpr mdarray(mapping_type const& m, container_policy_type const& cp)
```

Comment:
Since we are having our own implementations, we might consider padding here.  That way we will have 
`map_` (just for book keeping) and `padded_map_ `   and then `c_(cp_.create(padded_map_.required_span_size()))`

Reply:
Thank you for the suggestion @rg20 , yeah I think we will have a padded version of array and span by customization of layout.  I haven't added it in this PR yet.